[FLINK-24086][runtime] Rebuilding the SharedStateRegistry only when the restore method is called for the first time

rkhachatryan opened the pull request:
Copy of #17179 to rebase, address feedback and merge.

```
 [FLINK-24086][checkpoint] Rebuilding the SharedStateRegistry only when the restore method is called for the first time.
 
From FLINK-22483, the CompletedCheckpointStore will not change during task failover, so we only need to rebuild the SharedStateRegistry once, which can reduce the recovery time during failover.


<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Move `SharedStateRigistry` to `CompleteCheckpointStore` to make the life cycle of them consistent, so that we only need to re-register the shared state once when the `CompleteCheckpointStore` is restored, which can reduce the time of task failover.

## Brief change log

1. Add `SharedStateRegistry` to the `createRecoveredCompletedCheckpointStore` method of `CheckpointRecoveryFactory`, and let `CompletedCheckpointStore` manage `SharedStateRegistry` by itself. `CheckpointCoordinator` no longer manages `SharedStateRegistry`.
2. When the `CompletedCheckpointStore` is recovering, it will also register the shared state of the recovered `CompletedCheckpoint`.
3. `CompletedCheckpointStore` adds the `registerSharedState` method, which is used to provide an interface for the `CheckpointCoordinator` to register the shared state.

## Verifying this change

This change added tests and can be verified as follows:
  - *Added a test to verify that the shared state reference is correct and can be deleted correctly when the SharedStateRegistry is not rebuilt during failover.*
>  org.apache.flink.runtime.checkpoint.CheckpointCoordinatorTest#testSharedStateRegistrationWithoutRebuildSharedStateRegistry

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
```

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PerJobCheckpointRecoveryFactory.java:
```diff
@@ -38,7 +42,7 @@
             CheckpointRecoveryFactory withoutCheckpointStoreRecovery(IntFunction<T> storeFn) {
         return new PerJobCheckpointRecoveryFactory<>(
-                (maxCheckpoints, previous) -> {
+                (maxCheckpoints, previous, sharedStateRegistry, ioExecutor) -> {
```
```suggestion
                (maxCheckpoints, previous, sharedStateRegistryFactory, ioExecutor) -> {
```

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedHaServicesWithLeadershipControl.java:
```diff
@@ -36,16 +40,23 @@ public EmbeddedHaServicesWithLeadershipControl(Executor executor) {
+                                    previous != null
+                                            ? previous.getAllCheckpoints()
+                                            : Collections.emptyList();
+                            SharedStateRegistry stateRegistry =
```
This is something the original PR tried to change. The original PRs intention was not to recreate the `SharedStateRegistry` if we are not recreating the `CompletedCheckpointStore`.

rkhachatryan replied:
`EmbeddedCompletedCheckpointStore` **is** created a few lines below:
```
                            SharedStateRegistry stateRegistry =
                                    stateRegistryFactory.create(ioExecutor, checkpoints);
                            if (previous != null) {
                                if (!previous.getShutdownStatus().isPresent()) {
                                    throw new IllegalStateException(
                                            "Completed checkpoint store from previous run has not yet shutdown.");
                                }
                                return new EmbeddedCompletedCheckpointStore(
                                        maxCheckpoints, checkpoints, stateRegistry);
                            }
                            return new EmbeddedCompletedCheckpointStore(
                                    maxCheckpoints, checkpoints, stateRegistry);
```
Or do you mean something else?

dawidwys replied:
Just in terms of classes. In fact it is not recreated. IMO, we could even simply return `previous`. Not sure why we actually don't do it. 

The intention was not to recreate the `SharedStateRegistry` if we have `previous`, because the one in previous should already be populated with all of the completed checkpoints. Populating the `SharedStateRegistry` turns out to be an expensive operation that the user would prefer to omit.

rkhachatryan replied:
> Just in terms of classes. In fact it is not recreated

Sorry I don't get you: `EmbeddedCompletedCheckpointStore` object is recreated with `new` and it represents the store, doesn't it?
The checkpoints from the old store **are** reused, but they are immutable; the registry, OTH could be closed, so I think it would be error-prone to reuse it.

I think the intention is to not re-create the registry in `CheckpointCoordinator.restoreLatestCheckpointedStateInternal`;
(currently, it re-uses the store but re-creates the registry).

Btw, I see that `EmbeddedHaServicesWithLeadershipControl` is used only in tests.

dawidwys replied:
Yes, you are obviously right. I was talking bollocks. I got completely confused :( 

Sorry for the confusion.

rkhachatryan replied:
No worries, thanks for raising these concerns :slightly_smiling_face:

dawidwys reviewed (approved):
Looks good from my side.

rkhachatryan commented:
Thanks for the review @dawidwys !
I'll squash the commits and merge the PR.

dawidwys commented:
I think there is one more issue, I've just spotted. With the current changes, we won't register shared states if we restore from an externalized checkpoint (via `CheckpointCoordinator#restoreSavepoint`) which might lead to wrong reference counting. WDYT @rkhachatryan ?

rkhachatryan commented:
Good point @dawidwys ,
I've added registration in `restoreSavepoint` in 21c3929bce00ea7203bbbcfb5b4992797df570ac.

Previously, registration happened **after** adding a checkpoint to the store. 
I think this is also incorrect: adding to the store might trigger subsumption, so the ref counts must be up-to-date. So I added registration **before** adding to the store:
```
        savepoint.registerSharedStatesAfterRestored(
                completedCheckpointStore.getSharedStateRegistry());

        completedCheckpointStore.addCheckpoint(
                savepoint, checkpointsCleaner, this::scheduleTriggerRequest);
```

dawidwys commented:
Looks good! I have no further concerns.

dawidwys commented:
@flinkbot run azure

rkhachatryan commented:
Build failed because of unrelated FLINK-20928 but [succeeded](https://dev.azure.com/khachatryanroman/flink/_build/results?buildId=1260&view=logs&s=ae4f8708-9994-57d3-c2d7-b892156e7812&j=d543d572-9428-5803-a30c-e8e09bf70915) in private branch, merging.
